plugins/ch341a/fu-ch341a-cfi-device.c

```diff
@@ -0,0 +1,431 @@
+				   gsize bufsz,
+				   FuProgress *progress,
+				   GError **error)
+{
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(FU_DEVICE(self)));
+	guint8 buf[CH341A_PAYLOAD_SIZE] = {0x0};
+	g_autoptr(FuDeviceLocker) cslocker = NULL;
+	g_autoptr(GByteArray) blob = g_byte_array_new();
+	g_autoptr(GPtrArray) chunks = NULL;
+
+	/* enable chip */
+	cslocker = fu_cfi_device_chip_select_locker_new(FU_CFI_DEVICE(self), error);
+	if (cslocker == NULL)
+		return NULL;
+
+	/* read each block */
+	chunks = fu_chunk_array_new(NULL, bufsz + 0x4, 0x0, 0x0, CH341A_PAYLOAD_SIZE);
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, chunks->len);
+	fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_READ);
+
+	/* cmd, then 24 bit starting address */
+	fu_common_write_uint32(buf, 0x0, G_BIG_ENDIAN);
+	if (!fu_cfi_device_get_cmd(FU_CFI_DEVICE(self),
+				   FU_CFI_DEVICE_CMD_READ_DATA,
+				   &buf[0],
+				   error))
+		return NULL;
+	for (guint i = 0; i < chunks->len; i++) {
+		FuChunk *chk = g_ptr_array_index(chunks, i);
+
+		/* the first package has cmd and address info */
+		if (!fu_ch341a_device_spi_transfer(proxy, buf, sizeof(buf), error))
+			return NULL;
+		if (i == 0) {
+			g_byte_array_append(blob, buf + 0x4, fu_chunk_get_data_sz(chk) - 0x4);
+		} else {
+			g_byte_array_append(blob, buf + 0x0, fu_chunk_get_data_sz(chk));
+		}
+
+		/* done */
+		fu_progress_step_done(progress);
+	}
+
+	/* success */
+	return g_byte_array_free_to_bytes(g_steal_pointer(&blob));
+}
+
+static gboolean
+fu_ch341a_cfi_device_write_firmware(FuDevice *device,
+				    FuFirmware *firmware,
+				    FuProgress *progress,
+				    FwupdInstallFlags flags,
+				    GError **error)
+{
+	FuCh341aCfiDevice *self = FU_CH341A_CFI_DEVICE(device);
+	FuCh341aDevice *proxy = FU_CH341A_DEVICE(fu_device_get_proxy(FU_DEVICE(self)));
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) fw_verify = NULL;
+	g_autoptr(GPtrArray) pages = NULL;
+	g_autoptr(FuDeviceLocker) cslocker = NULL;
+	g_autoptr(FuDeviceLocker) locker = NULL;
+
+	/* open programmer */
+	locker = fu_device_locker_new(proxy, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_ERASE, 33);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 44);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_VERIFY, 35);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* erase */
+	if (!fu_ch341a_cfi_device_write_enable(self, error)) {
+		g_prefix_error(error, "failed to enable writes: ");
+		return FALSE;
+	}
+	if (!fu_ch341a_cfi_device_chip_erase(self, error)) {
+		g_prefix_error(error, "failed to erase: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* write each block */
+	pages = fu_chunk_array_new_from_bytes(fw, 0x0, 0x0, 0x100);
```

Comment:
The `0x100` value is the page size? A transfert size value specific to ch341a?

Reply:
Or to the CFI device?

Reply:
Agreed, I need to use fu_cfi_device_get_page_size() -- fixed!